cmd/nvidia_gpu/nvidia_gpu.go

```diff
@@ -41,22 +43,52 @@ var (
 	enableContainerGPUMetrics      = flag.Bool("enable-container-gpu-metrics", false, "If true, the device plugin will expose GPU metrics for containers with allocated GPU")
 	gpuMetricsPort                 = flag.Int("gpu-metrics-port", 2112, "POrt on which GPU metrics for containers are exposed")
 	gpuMetricsCollectionIntervalMs = flag.Int("gpu-metrics-collection-interval", 30000, "Colection interval (in milli seconds) for container GPU metrics")
+	gpuConfigFile                  = flag.String("gpu-config", "/etc/nvidia/gpu_config.json", "File with GPU configurations for device plugin")
 )
 
+func parseGPUConfig(gpuConfigFile string) (gpumanager.GPUConfig, error) {
+	var gpuConfig gpumanager.GPUConfig
+
+	gpuConfigContent, err := ioutil.ReadFile(gpuConfigFile)
+	if err != nil {
+		return gpuConfig, fmt.Errorf("unable to read gpu config file %s: %v", gpuConfigFile, err)
+	}
+
+	if err = json.Unmarshal(gpuConfigContent, &gpuConfig); err != nil {
+		return gpuConfig, fmt.Errorf("failed to parse GPU config file contents: %s, error: %v", gpuConfigContent, err)
+	}
+	return gpuConfig, nil
+}
+
 func main() {
 	flag.Parse()
 	glog.Infoln("device-plugin started")
 	mountPaths := []gpumanager.MountPath{
 		{HostPath: *hostPathPrefix, ContainerPath: *containerPathPrefix},
 		{HostPath: *hostVulkanICDPathPrefix, ContainerPath: *containerVulkanICDPathPrefix}}
 
-	ngm := gpumanager.NewNvidiaGPUManager(devDirectory, mountPaths)
+	var gpuConfig gpumanager.GPUConfig
+	if *gpuConfigFile != "" {
+		glog.Infof("Reading GPU config file: %s", *gpuConfigFile)
+		var err error
+		gpuConfig, err = parseGPUConfig(*gpuConfigFile)
+		if err != nil {
+			glog.Infof("Failed to parse GPU config file %s: %v", *gpuConfigFile, err)
+			glog.Infof("Falling back to default GPU config.")
+			gpuConfig = gpumanager.GPUConfig{}
+		}
+	}
+	glog.Infof("Using gpu config: %v", gpuConfig)
+	ngm := gpumanager.NewNvidiaGPUManager(devDirectory, mountPaths, gpuConfig)
+
 	// Retry until nvidiactl and nvidia-uvm are detected. This is required
 	// because Nvidia drivers may not be installed initially.
 	for {
 		err := ngm.CheckDevicePaths()
 		if err == nil {
-			break
+			if err = ngm.Start(); err == nil {
```

Comment:
This is needed because ngm.Start() calls Start() on mig manager, which can return an error if the GPUs are not partitioned yet. This essentially makes the device plugin wait until all the GPU partitions are created.